### PR TITLE
Duet AI advanced DoFn documentation lookup prompts

### PR DESCRIPTION
**Pull request by olehborysevych**
- Side inputs
- Additional outputs
- Batched DoFn

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: learning/prompts/documentation-lookup/29_advanced_side_inputs.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        Integer sideInputValue = c.sideInput(sideInput);
+        ...
+      }
+    }).withSideInputs(sideInput));
+```
+
+Side inputs can also be used to process windowed data. Apache Beam uses the windows for the main input elements to look up the appropriate windows for the side input elements. Apache Beam projects the main input’s window to the side input's window set and uses the side input from the resulting window. Main and side inputs might have identical or different windowing strategies.
+For example, if the main input `PCollection` is windowed into ten-minute windows, and the side input is windowed into one-hour windows, Apache Beam projects the main input window set and selects the value from the corresponding one-hour side input window. For more information, refer to [side inputs and windowing](https://beam.apache.org/documentation/programming-guide/#side-inputs-windowing).
+
+For common patterns and code samples in Apache Beam pipelines to implement side inputs, refer to [side input patterns](https://beam.apache.org/documentation/patterns/side-inputs/).
```
Comment (dariabezkorovaina) [inline learning/prompts/documentation-lookup/29_advanced_side_inputs.md:34]:
```suggestion
For common patterns and code samples to implement side inputs in Apache Beam pipelines, refer to [side input patterns](https://beam.apache.org/documentation/patterns/side-inputs/).
```

File: learning/prompts/documentation-lookup/30_advanced_additional_outputs.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+         out.get(additionalOutputTagString).output(word);
+       }
+       if (condition for additional integer output) {
+         // Emit element to additional integer output
+         out.get(additionalOutputTagIntegers).output(word.length());
+       }
+     }
+```
+
+In the Apache Beam Python SDK, you can implement additional outputs by invoking the `with_outputs()` method on the `ParDo` and specifying the expected tags for the multiple outputs. The method returns a `DoOutputsTuple` object, with the specified tags serving as attributes that provide `ParDo` with access to the corresponding output `PCollection`s. For a sample Apache Beam Python pipeline demonstrating a word count example with multiple outputs, refer to the [multiple output `ParDo`](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/cookbook/multiple_output_pardo.py) example in the Apache Beam GitHub repository.
```
Comment (damccorm) [inline learning/prompts/documentation-lookup/30_advanced_additional_outputs.md:56]: Can we inline this example similar to the Java example? I don't think the training can follow the link, so it wouldn't be able to formulate an example

File: learning/prompts/documentation-lookup/31_advanced_batched_dofn.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+```python
+class MultiplyByTen(beam.DoFn):
+  def process(self, element: np.int64) -> Iterator[np.int64]:
+    yield element * 10
+
+  def process_batch(self, batch: np.ndarray) -> Iterator[np.ndarray]:
+    yield batch * 10
+```
+In this example, the ‘np.int64’ type represents the individual element. The 'process' method multiplies an element by ten, yielding a single element. The ‘np.ndarray’ type represents the batch. The 'process_batch' method multiplies each element in the batch by ten, yielding a single batch. During pipeline execution, Apache Beam will automatically select the best implementation based on the context.
```
Comment (damccorm) [inline learning/prompts/documentation-lookup/31_advanced_batched_dofn.md:25]:
```suggestion

In this example, the ‘np.int64’ type represents the individual element. The 'process' method multiplies an element by ten, yielding a single element. The ‘np.ndarray’ type represents the batch. The 'process_batch' method multiplies each element in the batch by ten, yielding a single batch. During pipeline execution, Apache Beam will automatically select the best implementation based on the context. 
```

nit

File: learning/prompts/documentation-lookup/31_advanced_batched_dofn.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+Here’s an example with chained ‘DoFn’s of equivalent types:
+
+```python
+(p | beam.Create([1, 2, 3, 4]).with_output_types(np.int64)
+   | beam.ParDo(MultiplyByTen()) # Implicit buffering and batch creation
+   | beam.ParDo(MultiplyByTen()) # Batches passed through
+   | beam.ParDo(MultiplyByTen()))
+```
+In this example, the ‘PTransform.with_output_types’ method sets the element-wise typehint for the output. Thus, when the `MultiplyByTen` class is applied to a `PCollection`, Apache Beam recognizes that `np.ndarray` is an acceptable batch type to use in conjunction with `np.int64` elements.
```
Comment (damccorm) [inline learning/prompts/documentation-lookup/31_advanced_batched_dofn.md:37]:
```suggestion

In this example, the ‘PTransform.with_output_types’ method sets the element-wise typehint for the output. Thus, when the `MultiplyByTen` class is applied to a `PCollection`, Apache Beam recognizes that `np.ndarray` is an acceptable batch type to use in conjunction with `np.int64` elements.
```

**Review by damccorm — COMMENTED** [review]
Mostly looks good, just had a few minor comments

**Comment by olehborysevych** [comment]
Hey @damccorm thank you for review. Your comment's are resolved.

**Review by damccorm — APPROVED** [review]
Thanks!

**Pull request merged**